### PR TITLE
Author Blog Articles plugin improvement

### PR DESCRIPTION
**Pull request by corentinbettiol**
*tl;dr: We want a "**get the N latest posts written by this author**" feature.*

## Context
Currently, here's the look of the "Author Blog Articles" plugin:

image

There's two problems here:
1. Each author is displayed once for each article he or she has published.
  https://github.com/nephila/djangocms-blog/blob/b10cbda6528527b8e25813b41f890c604ada8330/djangocms_blog/models.py#L535
  *This is a well-known, [old django bug](https://code.djangoproject.com/ticket/11707).*
2. There is a "*Articles*" input. "*The number of author articles to be displayed*" is not really used (*but it could be*).

image

See? We have set the value "Articles" to 1, but all the plugin does is to display the first and last names of the selected authors, as well as the number of blog posts they have written.

## What is done?
*We have looked in the code of djangocms-blog.*

Here we are in `models.py`:

image

There is two things to notice here:
1. `AuthorEntriesPlugin` have a `get_posts` function (*but that function is not used*),
2. We are getting a queryset of the blog posts posted by each author selected in the plugin (with `djangocms_blog_post_author`, a [`RelatedManager`](https://docs.djangoproject.com/en/3.0/ref/models/relations/#django.db.models.fields.related.RelatedManager)), we count them, and then we add this number to the [context](https://user-images.githubusercontent.com/45763865/70983935-d9014880-20b9-11ea-8c14-4954e05e3b70.png) (in `cms_plugins.py`) so the template can display this number:

image

image

## What we want?

We want to add the posts in the context in addition to the total number of articles posted (the `count`).
This can be done by slicing the queryset with `self.latest_posts` (the *Articles - number of author articles to be displayed* value), and adding this queryset to the `authors` 

image

## Results:

 * Here's a view of the blog used to test the feature:
  image
 * Here's the updated template used to test the feature (*[link to a gist](https://gist.github.com/corentinbettiol/d4e028938918224ade665e44d5d43e4f)*):
  image
 * And here are the results:
  image
  image

### CODE REVIEW

**Comment by yakky** [comment]
Wow, thanks for the super detailed explanation
I agree it could be useful to report the author posts in an attribute

I am not sure the two feature should be in the same plugin as:
- they require different templates
- different plugins will communicate better the behavior

I will think about the best approach to this and will report soon

**Comment by yakky** [comment]
Alternatively the same result can be achieved entirely in the template to avoid the costant performance hit of the extra queryset even if when the post list is not used

**Comment by corentinbettiol** [comment]
We though that adding the posts in a plugin where the [`__str__` function](https://github.com/nephila/djangocms-blog/blob/b10cbda6528527b8e25813b41f890c604ada8330/djangocms_blog/models.py#L543) contain "*number* latest articles by author" would be the right thing to do.

I don't know why it was not implemented 5 years ago, but the class even have a `get_posts` function (that do [not work](https://user-images.githubusercontent.com/45763865/71070607-6b1d5580-217b-11ea-8cf6-687816d73d77.png) the way I though it would work).

(*[here are the modifications](https://user-images.githubusercontent.com/45763865/71070728-ae77c400-217b-11ea-966d-6acfdce23145.png) made to make the previous screenshot, I just added a new context var, containing the result of the `get_posts` function*)

I think that the "Author Blog Articles" plugin should display the Nth last posts per selected author, and that the display of the number of published articles should instead be added in a new plugin ("Author Articles Number"?).

**Comment by corentinbettiol** [comment]
(how do we have access to `latest_posts` to limit the queryset in the template ?)

**Comment by yakky** [comment]
> (how do we have access to `latest_posts` to limit the queryset in the template ?)

As `instance` contains the plugin model instance you can do `{{ instance.latest_post }}`

**Comment by yakky** [comment]
> > (how do we have access to `latest_posts` to limit the queryset in the template ?)
> 
> As `instance` contains the plugin model instance you can do `{{ instance.latest_post }}`

Th

> Alternatively the same result can be achieved entirely in the template to avoid the costant performance hit of the extra queryset even if when the post list is not used

I retract this comment as it's likely not doable as we can't pass to `slice` a context variable

**Comment by yakky** [comment]
@corentinbettiol based on your great work and investigation I extended this PR to make a few tweaks / additions over this.
You great analysis and initial work helped me a lot focusing on the issues of authors plugin

**Comment by corentinbettiol** [comment]
Thanks for the feedback and for the PR :)

I spent some time understanding the structure of the code, and I wanted the "workflow" to be explained in depth in at least one place (so that I could figure it out later, and so that other potential contributors wouldn't have to search through the code).

**Comment by yakky** [comment]
@corentinbettiol we will introduce some code refactoring and documentation after 1.1 to make easier to understand the code structure
If you don't have further objects I will merge #533 to release 1.1 this coming week

**Comment by corentinbettiol** [comment]
That's good for me, thanks for notifying me of that :)
